README.md
# Toggle Text Wrapping, a plugin for gedit #

Quickly toggle text wrapping from the View menu or Toolbar  
<https://github.com/jefferyto/gedit-toggle-text-wrapping>  
0.3.2

All bug reports, feature requests and miscellaneous comments are welcome
at the [project issue tracker][].

## Requirements ##

0.3.0 and higher requires gedit 3; gedit 2 users should use [0.2.0][] by
Christian Hartmann.

## Installation ##

1.  Download the source code (as [zip][] or [tar.gz][]) and extract.
2.  Copy the `toggletextwrapping` folder and the appropriate `.plugin`
    file into `~/.local/share/gedit/plugins` (create if it does not
    exist):
    *   For gedit 3.6 and earlier, copy
        `toggletextwrapping.plugin.python2` and rename to
        `toggletextwrapping.plugin`.
    *   For gedit 3.8 and later, copy `toggletextwrapping.plugin`.
3.  Restart gedit, select **Edit > Preferences** (or
    **gedit > Preferences** on Mac), and enable the plugin in the
    **Plugins** tab.

## Usage ##

To toggle text wrapping for the current document, either:

*   Select **View > Enable Text Wrapping**;
*   Click the checkmark button in the Toolbar; or
*   Use the <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>B</kbd> keyboard
    shortcut.

Note that this only affects the current (active) document; text wrapping
for documents in other tabs and/or windows will be unaffected, and the
global text wrapping preferences will remain unchanged.

## Development ##

The code in `toggletextwrapping/utils` comes from [python-gtk-utils][];
changes should ideally be contributed to that project, then pulled back
into this one with `git subtree pull`.

## Credits ##

Christian Hartmann:  
Parts of this plugin are based on the work of Mike Doty <mike@psyguygames.com>
who wrote the infamous SplitView plugin. The rest is inspired from the Python
Plugin Howto document and the Python-GTK documentation.

Francisco Franchetti:  
This plugin was developed for gedit 2 by Christian Hartmann at <christian.hartmann@berlin.de>.

Jeffery To:  
Based on the work by Christian Hartmann and Francisco Franchetti :-)

## License ##

Copyright &copy; 2008-2009 Christian Hartmann <christian.hartmann@berlin.de>  
Copyright &copy; 2011 Francisco Franchetti <nixahn@gmail.com>  
Copyright &copy; 2013 Jeffery To <jeffery.to@gmail.com>

Available under GNU General Public License version 3


[project issue tracker]: https://github.com/jefferyto/gedit-toggle-text-wrapping/issues
[zip]: https://github.com/jefferyto/gedit-toggle-text-wrapping/archive/master.zip
[tar.gz]: https://github.com/jefferyto/gedit-toggle-text-wrapping/archive/master.tar.gz
[0.2.0]: http://hartmann-it-design.de/gedit/TextWrap/index.html
[python-gtk-utils]: https://github.com/jefferyto/python-gtk-utils


Changelog
2013-11-01  Jeffery To  <jeffery.to@gmail.com>

	0.3.2:

	* Added a toggle button to the fullscreen toolbar

2013-10-31  Jeffery To  <jeffery.to@gmail.com>

	0.3.1, a "minor" release:

	* When toggling text wrapping, use the user's preferred wrap mode
	  (from preferences) when possible
	* Fixed toggle state when the user changes their text wrapping
	  preferences (and thus changing the wrap mode of all open
	  documents)
	* Made .plugin file compatible with gedit 3.8, adding a separate
	  .plugin file for gedit < 3.8
	* Code cleanup

2011-11-05  Francisco Franchetti  <nixahn@gmail.com>

	migration to gedit 3 by Francisco Franchetti. things changed are
	in the .py file

	_ dont use the enums in capital letters; look up the numbers online and use those
	_ gtk -> Gtk
	_ add the do_ in front of activate, deactivate, update_state (not update_ui); a good practice would be to add it to the custom methods too
	_ def statements don't have the window argument (defined for the class in the beginning)
	_ import statements are different
	_ class definition, init are different
	_ delete first line with the !/dev...
	in the .plugin file
	_ [Gedit plugin] -> [Plugin]
	_ IAge=3

2009-04-26  Christian Hartmann  <christian.hartmann@berlin.de>

	0.2 changed filenames from textwrap to TextWrap as it conflicts with
	/usr/lib/python2.6/textwrap.py when loading the plugin. Unfortunately
	i have no real clue what actualy is causing this conflict. This might
	be reasoned by a change in the Gedit Python Plugin Loader, as this has
	not been happening before upgrading gedit or a prerequisite of it through
	an upgrade of my Ubuntu to 8.10 or 9.04. Added a couple documentst mainly
	to ease the burdon of installation for gedit plugin beginners and made it
	public available on my company website: http://hartmann-it-design.de/gedit

2008-10-10  Christian Hartmann  <christian.hartmann@berlin.de>

	0.1 initial release for private use only


toggletextwrapping.py
# -*- coding: utf8 -*-
#
# toggletextwrapping.py
# This file is part of Toggle Text Wrapping, a plugin for gedit
#
# Copyright (C) 2008-2009 Christian Hartmann <christian.hartmann@berlin.de>
# Copyright (C) 2011 Francisco Franchetti <nixahn@gmail.com>
# Copyright (C) 2013 Jeffery To <jeffery.to@gmail.com>
# https://github.com/jefferyto/gedit-toggle-text-wrapping
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from gi.repository import GObject, Gtk, Gedit

ui_str = """<ui>
	<menubar name="MenuBar">
		<menu name="ViewMenu" action="View">
			<placeholder name="ViewOps_2">
				<menuitem name="ToggleTextWrappingMenuItem" action="ToggleTextWrappingPluginToggle" />
			</placeholder>
		</menu>
	</menubar>
	<toolbar name="ToolBar">
		<separator />
		<toolitem name="ToggleTextWrappingToolItem" action="ToggleTextWrappingPluginToggle" />
	</toolbar>
</ui>
"""

class ToggleTextWrappingPlugin(GObject.Object, Gedit.WindowActivatable):
	__gtype_name__ = 'ToggleTextWrappingPlugin'

	window = GObject.property(type=Gedit.Window)

	TOGGLE_ACCELERATOR = '<Ctrl><Shift>B'

	DEFAULT_WRAP_MODE = Gtk.WrapMode.WORD # or Gtk.WrapMode.CHAR

	HANDLER_IDS = 'ToggleTextWrappingPluginHandlerIds'

	def __init__(self):
		GObject.Object.__init__(self)

	def do_activate(self):
		action_group = Gtk.ActionGroup('ToggleTextWrappingPluginActions')
		action_group.set_translation_domain('gedit')
		action = Gtk.ToggleAction('ToggleTextWrappingPluginToggle', _("Enable Text Wrapping"), _("Toggle text wrapping for the current document"), Gtk.STOCK_OK)
		action_group.add_action_with_accel(action, self.TOGGLE_ACCELERATOR)

		self._connect_handlers(action, ('activate',), 'toggle_action')

		manager = self.window.get_ui_manager()
		manager.insert_action_group(action_group, -1)
		ui_id = manager.add_ui_from_string(ui_str)

		self._ui_id = ui_id
		self._action = action
		self._action_group = action_group

		self.do_update_state()

	def do_deactivate(self):
		self._disconnect_handlers(self._action)

		manager = self.window.get_ui_manager()
		manager.remove_ui(self._ui_id)
		manager.remove_action_group(self._action_group)
		manager.ensure_update()

		self._ui_id = None
		self._action = None
		self._action_group = None

	def do_update_state(self):
		view = self.window.get_active_view()
		self._action.set_sensitive(view is not None)
		self._action.set_active(view is not None and view.get_wrap_mode() != Gtk.WrapMode.NONE)

	def on_toggle_action_activate(self, action):
		view = self.window.get_active_view()
		if view:
			view.set_wrap_mode(self.DEFAULT_WRAP_MODE if action.get_active() else Gtk.WrapMode.NONE)

	def _connect_handlers(self, obj, signals, m, *args):
		HANDLER_IDS = self.HANDLER_IDS
		l_ids = getattr(obj, HANDLER_IDS) if hasattr(obj, HANDLER_IDS) else []

		for signal in signals:
			if type(m).__name__ == 'str':
				method = getattr(self, 'on_' + m + '_' + signal.replace('-', '_').replace('::', '_'))
			else:
				method = m
			l_ids.append(obj.connect(signal, method, *args))

		setattr(obj, HANDLER_IDS, l_ids)

	def _disconnect_handlers(self, obj):
		HANDLER_IDS = self.HANDLER_IDS
		if hasattr(obj, HANDLER_IDS):
			for l_id in getattr(obj, HANDLER_IDS):
				obj.disconnect(l_id)

			delattr(obj, HANDLER_IDS)

	def _block_handlers(self, obj):
		HANDLER_IDS = self.HANDLER_IDS
		if hasattr(obj, HANDLER_IDS):
			for l_id in getattr(obj, HANDLER_IDS):
				obj.handler_block(l_id)

	def _unblock_handlers(self, obj):
		HANDLER_IDS = self.HANDLER_IDS
		if hasattr(obj, HANDLER_IDS):
			for l_id in getattr(obj, HANDLER_IDS):
				obj.handler_unblock(l_id)


toggletextwrapping/util.py
# -*- coding: utf-8 -*-
#
# util.py
# This file is part of Toggle Text Wrapping, a plugin for gedit
#
# Copyright (C) 2013 Jeffery To <jeffery.to@gmail.com>
# https://github.com/jefferyto/gedit-toggle-text-wrapping
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

def _get_handler_ids_name(ns):
	return ns.__class__.__name__ + 'HandlerIds'

def _get_handler_ids(ns, target):
	name = _get_handler_ids_name(ns)
	return getattr(target, name) if hasattr(target, name) else []

def _set_handler_ids(ns, target, ids):
	name = _get_handler_ids_name(ns)
	setattr(target, name, ids)

def _del_handler_ids(ns, target):
	name = _get_handler_ids_name(ns)
	if hasattr(target, name):
		delattr(target, name)

def connect_handlers(ns, target, signals, prefix_or_fn, *args):
	handler_ids = _get_handler_ids(ns, target)

	for signal in signals:
		if hasattr(prefix_or_fn, '__call__'):
			fn = prefix_or_fn
		else:
			fn = getattr(ns, 'on_%s_%s' % (prefix_or_fn, signal.replace('-', '_').replace('::', '_')))
		handler_ids.append(target.connect(signal, fn, *args))

	_set_handler_ids(ns, target, handler_ids)

def disconnect_handlers(ns, target):
	for handler_id in _get_handler_ids(ns, target):
		target.disconnect(handler_id)

	_del_handler_ids(ns, target)

def block_handlers(ns, target):
	for handler_id in _get_handler_ids(ns, target):
		target.handler_block(handler_id)

def unblock_handlers(ns, target):
	for handler_id in _get_handler_ids(ns, target):
		target.handler_unblock(handler_id)


toggletextwrapping/__init__.py
# -*- coding: utf-8 -*-
#
# __init__.py
# This file is part of Toggle Text Wrapping, a plugin for gedit
#
# Copyright (C) 2008-2009 Christian Hartmann <christian.hartmann@berlin.de>
# Copyright (C) 2011 Francisco Franchetti <nixahn@gmail.com>
# Copyright (C) 2013 Jeffery To <jeffery.to@gmail.com>
# https://github.com/jefferyto/gedit-toggle-text-wrapping
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from gi.repository import GObject, Gtk, Gio, Gedit
from .utils import connect_handlers, disconnect_handlers, block_handlers, unblock_handlers

ui_str = """<ui>
	<menubar name="MenuBar">
		<menu name="ViewMenu" action="View">
			<placeholder name="ViewOps_2">
				<menuitem name="ToggleTextWrappingMenuItem" action="ToggleTextWrappingPluginToggle" />
			</placeholder>
		</menu>
	</menubar>
	<toolbar name="ToolBar">
		<separator />
		<toolitem name="ToggleTextWrappingToolItem" action="ToggleTextWrappingPluginToggle" />
	</toolbar>
</ui>
"""

class ToggleTextWrappingPlugin(GObject.Object, Gedit.WindowActivatable):
	__gtype_name__ = 'ToggleTextWrappingPlugin'

	window = GObject.property(type=Gedit.Window)

	TOGGLE_ACCELERATOR = '<Ctrl><Shift>B'

	DEFAULT_WRAP_MODE = Gtk.WrapMode.WORD # or Gtk.WrapMode.CHAR

	WRAP_MODE_SETTINGS_SCHEMA = 'org.gnome.gedit.preferences.editor'
	WRAP_MODE_SETTINGS_KEY = 'wrap-mode'

	def __init__(self):
		GObject.Object.__init__(self)

	def do_activate(self):
		window = self.window

		action_group = Gtk.ActionGroup('ToggleTextWrappingPluginActions')
		action_group.set_translation_domain('gedit')
		action = Gtk.ToggleAction('ToggleTextWrappingPluginToggle', _("Enable Text Wrapping"), _("Toggle text wrapping for the current document"), Gtk.STOCK_OK)
		action_group.add_action_with_accel(action, self.TOGGLE_ACCELERATOR)

		connect_handlers(self, action, ('activate',), 'toggle_action')

		manager = window.get_ui_manager()
		manager.insert_action_group(action_group, -1)
		ui_id = manager.add_ui_from_string(ui_str)

		fullscreen_ui_id = manager.new_merge_id()
		manager.add_ui(fullscreen_ui_id, '/FullscreenToolBar/SearchReplace', 'ToggleTextWrappingFullscreenToolItem', 'ToggleTextWrappingPluginToggle', Gtk.UIManagerItemType.TOOLITEM, False)
		manager.add_ui(fullscreen_ui_id, '/FullscreenToolBar/SearchReplace', '', None, Gtk.UIManagerItemType.SEPARATOR, False)

		settings = Gio.Settings.new(self.WRAP_MODE_SETTINGS_SCHEMA)
		connect_handlers(self, settings, ('changed::' + self.WRAP_MODE_SETTINGS_KEY,), 'pref')

		self._ui_id = ui_id
		self._fullscreen_ui_id = fullscreen_ui_id
		self._action = action
		self._action_group = action_group
		self._settings = settings
		self._wrap_mode = self.DEFAULT_WRAP_MODE

		self.on_pref_changed_wrap_mode(settings, self.WRAP_MODE_SETTINGS_KEY)

		for doc in window.get_documents():
			self.on_window_tab_added(window, Gedit.Tab.get_from_document(doc))

		connect_handlers(self, window, ('tab-added', 'tab-removed'), 'window')

		self.do_update_state()

	def do_deactivate(self):
		window = self.window

		disconnect_handlers(self, window)

		for doc in window.get_documents():
			self.on_window_tab_removed(window, Gedit.Tab.get_from_document(doc))

		disconnect_handlers(self, self._settings)
		disconnect_handlers(self, self._action)

		manager = window.get_ui_manager()
		manager.remove_ui(self._fullscreen_ui_id)
		manager.remove_ui(self._ui_id)
		manager.remove_action_group(self._action_group)
		manager.ensure_update()

		self._ui_id = None
		self._fullscreen_ui_id = None
		self._action = None
		self._action_group = None
		self._settings = None
		self._wrap_mode = None

	def do_update_state(self):
		view = self.window.get_active_view()
		action = self._action
		action.set_sensitive(view is not None)
		action.set_active(view is not None and view.get_wrap_mode() != Gtk.WrapMode.NONE)

	def on_toggle_action_activate(self, action):
		view = self.window.get_active_view()
		if view:
			block_handlers(self, view)
			view.set_wrap_mode(self._wrap_mode if action.get_active() else Gtk.WrapMode.NONE)
			unblock_handlers(self, view)

	def on_pref_changed_wrap_mode(self, settings, key):
		value = settings.get_string(key)
		if value == 'word':
			self._wrap_mode = Gtk.WrapMode.WORD
		elif value == 'char':
			self._wrap_mode = Gtk.WrapMode.CHAR

	def on_window_tab_added(self, window, tab):
		connect_handlers(self, tab.get_view(), ('notify::wrap-mode',), 'view')

	def on_window_tab_removed(self, window, tab):
		disconnect_handlers(self, tab.get_view())

	def on_view_notify_wrap_mode(self, view, prop):
		if self.window.get_active_view() == view:
			self.do_update_state()


toggletextwrap.py
# -*- coding: utf8 -*-
# Text Wrap Gedit Plugin
#
# This file is part of the Text Wrap Plugin for Gedit
# Copyright (C) 2008-2009 Christian Hartmann <christian.hartmann@berlin.de>
# Copyright (C) 2011 Francisco Franchetti <nixahn@gmail.com>
# Copyright (C) 2013 Jeffery To <jeffery.to@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# This plugin is intended to ease the setting of Text Wrap (aka Line Wrap,
# Word Wrap) by either a Keyboard Shortcurt (currently sticked to Shift-Ctrl-B),
# a new entry in the View Menu or by an Icon in the Toolbar. The use of either 
# option works as a toggle (de- or activate text wrap). The initial setting for 
# new or new opened files is taken from the setting in the Preferences dialog 
# and remembered per file as long thew file is open. 

# This plugin was developed for gedit 2 by Christian Hartmann at <christian.hartmann@berlin.de>. Parts of this plugin are based on the work of Mike Doty <mike@psyguygames.com>
# who wrote the infamous SplitView plugin. The rest is inspired from the Python
# Plugin Howto document and the Python-GTK documentation.

# CHANGELOG
# =========
# * 2008-10-10:
#   0.1 initial release for private use only
# * 2009-04-26:
#   0.2 changed filenames from textwrap to TextWrap as it conflicts with 
#   /usr/lib/python2.6/textwrap.py when loading the plugin. Unfortunately
#   i have no real clue what actualy is causing this conflict. This might
#   be reasoned by a change in the Gedit Python Plugin Loader, as this has
#   not been happening before upgrading gedit or a prerequisite of it through
#   an upgrade of my Ubuntu to 8.10 or 9.04. Added a couple documentst mainly
#   to ease the burdon of installation for gedit plugin beginners and made it
#   public available on my company website: http://hartmann-it-design.de/gedit
# * 2011-11-05:
#   migration to gedit 3 by Francisco Franchetti. things changed are
#   in the .py file
#   _ dont use the enums in capital letters; look up the numbers online and use those
#   _ gtk -> Gtk
#	_ add the do_ in front of activate, deactivate, update_state (not update_ui); a good practice would be to add it to the custom methods too
#   _ def statements don't have the window argument (defined for the class in the beginning)
#   _ import statements are different
#	_ class definition, init are different
#   _ delete first line with the !/dev...
#   in the .plugin file
#   _ [Gedit plugin] -> [Plugin]
#   _ IAge=3

# import basic requisites
from gi.repository import GObject, Gtk, Gdk, Gedit

# just a constant used in several places herein
prefix = "Plugin TextWrap "

# a common ui definition for menu and toolbar additions
ui_str = """<ui>
  <menubar name="MenuBar">
	<menu name="ViewMenu" action="View">
	  <placeholder name="ViewOps_2">
		<menuitem name="ToggleTextWrap" action="TTextWrap" />
	  </placeholder>
	</menu>
  </menubar>
  <toolbar name="ToolBar">
	<separator />
	<toolitem name="ToggleTextWrap" action="TTextWrap" />
  </toolbar>
</ui>
"""

# define the plugin class (helper class is not needed anymore cause gtk3 has windowactivatable)
class ToggleTextWrap(GObject.Object,Gedit.WindowActivatable):
	__gtype_name__= "ToggleTextWrap"
	window=GObject.property(type=Gedit.Window)

	def __init__(self):
		GObject.Object.__init__(self)

	def do_toggle_text_wrap(self, action):
		view = self.window.get_active_view()
		current_action = self._action_group.get_action("TTextWrap")
		if current_action.get_active():
			view.set_wrap_mode(2)
		else:
			view.set_wrap_mode(0)

	def do_activate(self):
		
		# Get initial state from text wrapping in this view (not available
		# on gedit startup but if plugin is enabled during the gedit session
		# and for what ever reason we do not have an update ui signal on init)
		view = self.window.get_active_view()
		try:
			current_wrap_mode = view.get_wrap_mode()
			# the order gives the numbers, starting at 0 for WRAP_NONE
			# typedef enum {
			# 	GTK_WRAP_NONE,
			# 	GTK_WRAP_CHAR,
			# 	GTK_WRAP_WORD,
			# 	GTK_WRAP_WORD_CHAR
			# } GtkWrapMode;
			if current_wrap_mode == 0:
				self._initial_toggle_state = False
			else:
				self._initial_toggle_state = True
		except:
			# Define default initial state for the plugin (should read this from the preferences file)
			self._initial_toggle_state = False
			# view.set_wrap_mode(0)	

		# Add "Toggle Text Wrap" to the View menu and to the Toolbar
		# Get the GtkUIManager
		self._manager = self.window.get_ui_manager()
		# Create a new action group
		self._action_group = Gtk.ActionGroup("GeditTextWrapPluginActions")
		self._action_group.add_toggle_actions([(
				"TTextWrap", 
				"gtk-ok", 
				_("Text Wrap"), 
				"<Ctrl><Shift>B", 
				_("Toggle Current Text Wrap Setting"), 
				self.do_toggle_text_wrap,
				self._initial_toggle_state)])
		# Insert the action group
		self._manager.insert_action_group(self._action_group)
		# Add my item to the "Views" menu and to the Toolbar
		self._ui_id = self._manager.add_ui_from_string(ui_str)
		# Debug merged ui
		self._manager.ensure_update()

	def do_deactivate(self):
		# Remove the ui
		self._manager.remove_ui(self._ui_id)
		self._ui_id = None
		# Remove action group
		self._manager.remove_action_group(self._action_group)
		self._action_group = None
		# ensure that manager updates
		self._manager.ensure_update()

	def do_update_state(self):
		view = self.window.get_active_view()
		self._action_group.set_sensitive(self.window.get_active_document() != None)
		# self._action_group.set_sensitive(bool(view and view.get_editable()))
		try:
			# Get initial state from word wrapping in this view (if any)
			current_wrap_mode = view.get_wrap_mode()
			# Get our action and set state according to current wrap mode
			current_action = self._action_group.get_action("TTextWrap")
			if current_wrap_mode == 0:
				current_action.set_active(False)
			else:
				current_action.set_active(True)
		except:
			return

		

#	def _console(self, vartext):
#		if self._DEBUG:
#			print prefix, vartext	

